### ISSUE TITLE
sql: improve regproc and other regfoo OID types 

### ISSUE THREAD

**Issue by jordanlewis** [cockroachdb/cockroach#13567]
Our current support for OID types is not complete and insufficient for proper ORM compatibility. In postgres, `regproc` and other `regfoo` types have the special property of being treatable as both an `OID` integer and a string, depending on the context. It turns out that lots of ORMs use this flexible feature, so we'll need to support it.

For an example, check out these queries against `pg_type.typreceive`, a column of type `regproc`:

```
jordan=# select typreceive from pg_type limit 1;
 typreceive
------------
 boolrecv

jordan=# select typreceive, proname from pg_type join pg_proc on pg_type.typreceive=pg_proc.oid limit 1;
 typreceive | proname
------------+----------
 boolrecv   | boolrecv
```

As you can see, `regproc` types are serialized as strings when rendered, but can be joined against `OID` columns.

Currently, we store all `OID` types as integers internally. We'll need to replace that representation with something like `type dOid struct { oid int64, name string, variant oidTypes }` instead, where the variant indicates what kind of `regfoo` the type is. This will allow the types to be serialized as a string, but properly treated as an integer/`oid` when necessary.

This will also allow us to get rid of the old `OID` datum type, replacing it with a `dOid` whose `variant` is something like `standard_oid` which wouldn't have that special string serialization rule.

Required for supporting both #9955 and #12526, as well as full support for ActiveRecord.

**Comment by knz** [comment]
This is addressed by #13896 right?

**Comment by jordanlewis** [comment]
Almost. There's a final missing piece, which is the behavior when casting integers or raw OIDs to reg* types. This kind of cast needs to do an query against the relevant table using the integer OID as the key. This is required by a few ORM queries.

I'll have a PR out for that today or tomorrow, at which point we can close this.
